inference_client_webrtc.py

```diff
@@ -0,0 +1,244 @@
+                self.in_buffer = self.in_buffer[BLOCK_SIZE:]
+            else:
+                audio_data = None
+            
+            # Process audio if available and add resulting audio to output buffer
+            if audio_data is not None:
+                if not self.sound_check:
+                    audio_data = await self.process_audio(audio_data)
+                if self.out_buffer is None:
+                    self.out_buffer = audio_data
+                else:
+                    self.out_buffer = np.concatenate((self.out_buffer, audio_data), axis=-1)
+
+            # Take `out_samples` samples from output buffer if available for output
+            out_samples = int(frame.samples * SAMPLE_RATE / STREAMING_SAMPLE_RATE)
+            if self.out_buffer is not None and self.out_buffer.shape[0] >= out_samples:
+                audio_data = self.out_buffer[:out_samples]
+                self.out_buffer = self.out_buffer[out_samples:]
+            else:
+                audio_data = None
+
+            # Output silence if no audio data available
+            if audio_data is None:
+                # output silence
+                audio_data = np.zeros(out_samples, dtype=np.int16)
+            
+            # Upsample output audio, convert to original number of channels, and convert to s16 format
+            audio_data = self.upsample(audio_data)
+            audio_data = self.to_channels(audio_data, len(frame.layout.channels))
+            audio_data = self.to_s16_format(audio_data)
+
+            # return audio data as AudioFrame
+            new_frame = av.AudioFrame.from_ndarray(audio_data, format=frame.format.name, layout=frame.layout.name)
+            new_frame.sample_rate = frame.sample_rate
+            out_frames.append(new_frame)
+
+        return out_frames
+
+    def stop(self): 
+        if self.ws is not None:
+            # TODO: this hangs. Figure out why.
+            #asyncio.get_event_loop().run_until_complete(self.ws.close())
+            print("Websocket closed")
+        self.ws = None
+        self.in_buffer = None
+        self.out_buffer = None
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description='Audio Client with Temperature Control')
+    parser.add_argument('--token_temp', '-t1', type=float, help='Token (LM) temperature parameter')
+    parser.add_argument('--categorical_temp', '-t2', type=float, help='Categorical (VAE) temperature parameter')
+    parser.add_argument('--gaussian_temp', '-t3', type=float, help='Gaussian (VAE) temperature parameter')
+    parser.add_argument('--server', '-s', default="ws://localhost:8000", 
+                        help='Server URL (default: ws://localhost:8000)')
+    parser.add_argument("--use_ice_servers", action="store_true", help="Use public STUN servers")
+    
+    args = parser.parse_args()
+    
+    # Audio settings
+    STREAMING_SAMPLE_RATE = 48000
+    SAMPLE_RATE = 16000
+    BLOCK_SIZE = 2000
+    CHANNELS = 1
+    
```

Comment:
```suggestion
    st.title("hertz-dev webrtc demo!")
    st.markdown("""
    Welcome to the audio processing interface! Here you can talk live with hertz.
    - Process audio in real-time through your microphone
    - Adjust various temperature parameters for inference
    - Test your microphone with sound check mode
    - Enable/disable echo cancellation and noise suppression
    
    To begin, click the START button below and allow microphone access.
    """)
```

Reply:
Might want to add some text to the main page—but was able to get the WebRTC live inference working locally for me! Had to pipe it through localhost `ssh -L 8501:localhost:8501 remote` but works!

Reply:
@neelr good suggestions, I'll take care of it shortly.